lib/charms/operator_libs_linux/v0/juju_systemd_notices.py

```diff
@@ -0,0 +1,485 @@
+    """Convert the specified name into an org.freedesktop.systemd1.Unit path handle.
+
+    Args:
+        name: The name of the service.
+
+    Returns:
+        String containing the DBus path.
+    """
+    # DBUS Object names may only contain ASCII chars [A-Z][a-z][0-9]_
+    # It's basically urlencoded but instead of a %, it uses a _
+    path = name
+    for key, value in _DBUS_CHAR_MAPPINGS.items():
+        path = path.replace(value, key)
+
+    return f"/org/freedesktop/systemd1/unit/{path}"
+
+
+def _dbus_path_to_name(path: str) -> str:
+    """Convert the specified DBus path handle to a service name.
+
+    Args:
+        path: The DBus path to convert to service name.
+
+    Returns:
+        String containing the service name.
+    """
+    # DBUS Object names may only contain ASCII chars [A-Z][a-z][0-9]_
+    name = Path(path).name
+    for key, value in _DBUS_CHAR_MAPPINGS.items():
+        name = name.replace(key, value)
+
+    return name
+
+
+def _systemd_unit_changed(msg: Message) -> bool:
+    """Send Juju notification if systemd unit state changes on the DBus bus.
+
+    Invoked when a PropertiesChanged event occurs on an org.freedesktop.systemd1.Unit
+    object across the dbus. These events are sent whenever a unit changes state, including
+    starting and stopping.
+
+    Args:
+        msg: The message to process in the callback.
+
+    Returns:
+        True if the event is processed. False if otherwise.
+    """
+    _logger.debug(
+        f"Received message: path: {msg.path}, interface: {msg.interface}, member: {msg.member}"
+    )
+    service = _dbus_path_to_name(msg.path)
+    properties = msg.body[1]
+    if "ActiveState" not in properties:
+        return False
+
+    global _service_states
+    if service not in _service_states:
+        _logger.debug(f"Dropping event for unwatched service: {service}")
+        return False
+
+    curr_state = properties["ActiveState"].value
+    prev_state = _service_states[service]
+    # Drop transitioning and duplicate events
+    if curr_state.endswith("ing") or curr_state == prev_state:
+        _logger.debug(f"Dropping event - service: {service}, state: {curr_state}")
+        return False
+
+    _service_states[service] = curr_state
+    _logger.debug(f"Service {service} changed state to {curr_state}")
+    # Run the hook in a separate thread so the dbus notifications aren't
+    # blocked from being received.
+    asyncio.create_task(_send_juju_notification(service, curr_state))
+    return True
+
+
+async def _send_juju_notification(service: str, state: str) -> None:
+    """Invoke a Juju hook to notify an operator that a service state has changed.
+
+    Args:
+        service: The name of the service which has changed state.
+        state: The state of the service.
+    """
+    if service.endswith(".service"):
+        service = service[0 : -len(".service")]
+    if state == "active":
+        event_name = "started"
+    else:
+        event_name = "stopped"
```

Comment:
Inline if/else is nice for this: `event_name = 'started' if state == 'active' else 'stopped'`

Reply:
Refactored in commit https://github.com/canonical/operator-libs-linux/pull/108/commits/a61cf98ddfaf54f73b3c56c9182208052ca01032